File: src/main/java/com/zippeykeys/praisebe/block/multi/MultiblockPart.java
package com.zippeykeys.praisebe.block.multi;

import com.zippeykeys.praisebe.factory.MultiblockPartBuilder;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import lombok.Getter;
import lombok.experimental.Wither;

import net.minecraft.block.state.IBlockState;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;


public class MultiblockPart implements Cloneable{
    @Getter
    protected BlockPos pos;

    @Getter
    @Wither
    protected final IBlockState blockState;

    @Getter
    @Wither
    protected final TileEntity tileEntity;

    public MultiblockPart(BlockPos posIn, IBlockState blockStateIn){
        this(posIn, blockStateIn, null);
    }

    public MultiblockPart(BlockPos posIn, IBlockState blockStateIn, @Nullable TileEntity tileEntityIn){
        pos = posIn;
        blockState = blockStateIn;
        tileEntity = tileEntityIn;
    }

    public boolean is(World world, BlockPos posIn){
        return world.getBlockState(pos) == blockState;
    }

    public void rotate(){
        pos = new BlockPos(-pos.getY(), pos.getX(), pos.getZ());
    }

    @Override
    public MultiblockPart clone(){
        return new MultiblockPart(pos, blockState, tileEntity);
    }

    @Contract(" -> new")
    public static MultiblockPartBuilder builder(){
        return new MultiblockPartBuilder();
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/block/multi/MultiblockPart.kt
package com.zippeykeys.praisebe.block.multi

import net.minecraft.block.state.IBlockState
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World


open class MultiblockPart constructor(pos: BlockPos,
        val blockState: IBlockState,
        val tileEntity: TileEntity? = null) : Cloneable {
    var pos: BlockPos
        protected set

    init {
        this.pos = pos
    }

    fun `is`(world: World, posIn: BlockPos): Boolean {
        return world.getBlockState(pos) === blockState
    }

    fun rotate() {
        pos = BlockPos(-pos.y, pos.x, pos.z)
    }

    public override fun clone(): MultiblockPart {
        return MultiblockPart(pos, blockState, tileEntity)
    }
}